### PR TITLE
Add sniff that will check that capabilities are used correctly.

### PR DESCRIPTION
**Pull request by dingo-d**
The sniff will check if the functions that are accepting capabilities as the argument actually use capabilities ([as described in the support article](https://wordpress.org/support/article/roles-and-capabilities/)) and not roles.

The sniff was rewritten from the original one in #1364 to include helper functions from PHPCSUtils. The list of core functions that are using capabilities as a parameter was updated with changes up to WordPress 6.1.0. The list of capabilities was updated as well. The sniff is also compatible with PHP 8 (named arguments).

Original PRs:
https://github.com/WordPress/WordPress-Coding-Standards/pull/1364
https://github.com/WPTT/WPThemeReview/pull/36

Ref:
https://make.wordpress.org/themes/handbook/review/required/#10-classic-themes
https://developer.wordpress.org/plugins/users/roles-and-capabilities/
https://developer.wordpress.org/plugins/security/checking-user-capabilities/

Closes: #1364

### CODE REVIEW

**Comment by jrfnl** [comment]
One remark: as part of working with @dingo-d on this sniff, I updated the list of known capabilities with anything I could find in Core. I would very much like to invite someone to very critically review that list as there is a risk that I may have added non-officially supported capabilities.

File: WordPress/Tests/WP/CapabilitiesUnitTest.inc
Comment (jrfnl) [inline WordPress/Tests/WP/CapabilitiesUnitTest.inc:17]:
```suggestion
	'super_admin' /* Comment */, // Error.
```

The `error` here is confusing as the `// Error` indicates the line should result in an error already. What is actually being checked is that a comment is ignored/disregarded, so let's just call it `/* Comment */`.

File: WordPress/Tests/WP/CapabilitiesUnitTest.inc
Comment (jrfnl) [inline WordPress/Tests/WP/CapabilitiesUnitTest.inc:28]: Nice one! I always forget there are people who code like this! 👍🏻

File: WordPress/Tests/WP/CapabilitiesUnitTest.inc
Comment (jrfnl) [inline WordPress/Tests/WP/CapabilitiesUnitTest.inc:88]: Noticed some things we don't have tests for. It's arbitrary whether or not to add them as they don't necessarily test the functionality of the sniff itself, but rather the functionality of the abstract sniff, but then again, the abstract sniff doesn't have its own tests, so until then, it kind of makes sense to have some tests for these...:
* Parse error/live coding, i.e. `add_posts_page( 'page_title',` (and nothing else - should be last test in the test file).
* `$obj->current_user_can()`
* `My\NS\add_posts_page(...)`
* etc

File: WordPress/Tests/WP/CapabilitiesUnitTest.php
Comment (jrfnl) [inline WordPress/Tests/WP/CapabilitiesUnitTest.php:86]: Suggestion: move this function to the top of the test file to be in line with other tests where we adjust these values.

**Review by jrfnl — CHANGES_REQUESTED** [review]
Looking great! Just some final touches and then we can get this merged (after ? years... ) 🎉

File: WordPress/Sniffs/WP/CapabilitiesSniff.php
Comment (GaryJones) [inline WordPress/Sniffs/WP/CapabilitiesSniff.php:22]:
```suggestion
 * User capabilities should be used, not roles or deprecated capabilities.
```

File: WordPress/Sniffs/WP/CapabilitiesSniff.php
Comment (GaryJones) [reply WordPress/Sniffs/WP/CapabilitiesSniff.php:321]: Should that second sentence be marked as an internal comment?

**Comment by dingo-d** [comment]
Blocked by: #2114. Once that PR is merged I will rebase and the tests should be passing.

File: WordPress/Sniffs/WP/CapabilitiesSniff.php
Comment (jrfnl) [inline WordPress/Sniffs/WP/CapabilitiesSniff.php:460]:
```suggestion
			'Found unknown capability "%s" in function call to %s(). Please check the spelling of the capability. If this is a custom capability, please verify the capability is registered with WordPress via a call to WP_Role(s)->add_cap().' . \PHP_EOL . 'Custom capabilities can be made known to this sniff by setting the "custom_capabilities" property in the PHPCS ruleset.',
```

Along the same lines as above. Trying to make the error message more actionable for end-users. Hoping this makes it clearer ?

File: WordPress/Sniffs/WP/CapabilitiesSniff.php
Comment (jrfnl) [inline WordPress/Sniffs/WP/CapabilitiesSniff.php:156]: Any particular reason why this property is `protected` ? 

```suggestion
	private $core_roles = array(
```

File: WordPress/Sniffs/WP/CapabilitiesSniff.php
Comment (jrfnl) [inline WordPress/Sniffs/WP/CapabilitiesSniff.php:180]: Any particular reason why this property is `protected` ? 

```suggestion
	private $core_capabilities = array(
```

File: WordPress/Sniffs/WP/CapabilitiesSniff.php
Comment (jrfnl) [inline WordPress/Sniffs/WP/CapabilitiesSniff.php:330]: Any particular reason why this property is `protected` ? 


```suggestion
	private $deprecated_capabilities = array(
```

File: WordPress/Sniffs/WP/CapabilitiesSniff.php
Comment (jrfnl) [inline WordPress/Sniffs/WP/CapabilitiesSniff.php:385]:
```suggestion
					array( $function_details['name'], $matched_content, $parameter['clean'] ),
```

Note: `$parameter['clean']` was unused without the `Found: ... ` addition.

**Review by jrfnl — CHANGES_REQUESTED** [review]
Thanks @dingo-d for working on this and making all those changes!

Went through it one last time with a fine toothcomb:
* Ran PHPCS with `--generator=text` to see the CLI docs output.
* Ran PHPCS against the test files to verify the error messages and codes.
* Re-ran PHPCS against the test files with a custom ruleset with the `<config name="minimum_supported_wp_version" value="2.9"/>` directive set.
* Re-ran PHPCS against the test files using the CLI `--runtime-set minimum_supported_wp_version 2.9` arg.
* Ran the tests.

These final remarks are mostly just small peanuts, dotting the i's and crossing the t's. The logic of the sniff is sound and doesn't need any changes anymore, the tests are good and comprehensive, the docs are clear. We're nearly there.

File: WordPress/Sniffs/WP/CapabilitiesSniff.php
Comment (dingo-d) [reply WordPress/Sniffs/WP/CapabilitiesSniff.php:411]: Maybe 

```suggestion
				'An empty string is not a valid capability. Empty string found as a parameter in a function call to %s()',
```

?

File: WordPress/Sniffs/WP/CapabilitiesSniff.php
Comment (dingo-d) [reply WordPress/Sniffs/WP/CapabilitiesSniff.php:330]: No reason, this is left over from the original PR 👍🏼

File: WordPress/Tests/WP/CapabilitiesUnitTest.php
Comment (dingo-d) [reply WordPress/Tests/WP/CapabilitiesUnitTest.php:31]: Yes, only the first file has the examples with the low-severity warnings. I'll add it to the condition list 👍🏼

File: WordPress/Sniffs/WP/CapabilitiesSniff.php
Comment (jrfnl) [reply WordPress/Sniffs/WP/CapabilitiesSniff.php:411]: I like where you're going with this. What about ?

```suggestion
				'An empty string is not a valid capability. Empty string found as the %s parameter in a function call to %s()"',
```
... and then adding `$function_details['name']` to  `$data` passed ?

**Review by jrfnl — APPROVED** [review]
> I would very much like to invite someone to very critically review that list as there is a risk that I may have added non-officially supported capabilities.

The review of the capabilities list is IMO not a blocker for merging, but would still be very welcome!

@sandeshjangam Would this ☝🏻 be something you'd be interested in helping us with ?

---

Other than the above remark from earlier on, everything looks to have been addressed.

Still saw three inconsistencies, so I've fixed those in a new commit (so you all can see easily what I changed). Feel free to squash on (or before) merge.

Approved (at long last). 🎉

**Review by GaryJones — APPROVED** [review]
✅

**Comment by dingo-d** [comment]
I agree with the added changes 👍🏼

**Comment by jrfnl** [comment]
Rebased without changes and squashed the extra commit into the original. Will merge once the build has passed.

**Pull request merged**
